resources/js/Components/UserActivityMap.tsx

```diff
@@ -0,0 +1,371 @@
+    }
+
+    /* array that holds cells for the table */
+    let tableData: JSX.Element[] = [];
+    let i;
+    let len = data.length;
+    const now = new Date();
+    let aggregateActivityTime = 0;
+
+    const convertSeconds = (secs: number) => {
+        let hours = Math.floor(secs / 3600);
+        let minutes = Math.floor((secs % 3600) / 60);
+        let seconds = Math.floor(secs % 60);
+        if (hours) {
+            return `${hours}h ${minutes}m ${seconds}s`;
+        } else if (minutes) {
+            return `${minutes}m ${seconds}s`;
+        } else {
+            return `${seconds}s`;
+        }
+    };
+
+    let dateCount = subtractYear(end);
+    if (end > now) {
+        end = now;
+    }
+
+    data.sort((a, b) => {
+        if (a.date < b.date) {
+            return -1;
+        } else if (a.date > b.date) {
+            return 1;
+        } else {
+            return 0;
+        }
+    });
+
+    /* add empty cells for days of week before activity range */
+    for (i = 0; i < dateCount.getUTCDay(); i++) {
+        tableData.push(
+            <td className="block" key={i}>
+                <div className={`${nodeSizes[size]}`}></div>
+            </td>,
+        );
+    }
+
+    i = 0;
+    /* fill cells with Nodes for days in range */
+    while (dateCount <= end) {
+        if (dateCount > now) {
+            tableData.push(
+                <td className="block" key={dateCount.getTime()}></td>,
+            );
+        } else if (
+            i < len &&
+            dateCount.toISOString().split("T")[0] == data[i].date
+        ) {
+            tableData.push(
+                <td className="block" key={dateCount.getTime()}>
+                    <Node
+                        date={new Date(dateCount)}
+                        activity_time={convertSeconds(
+                            Number(data[i].total_activity_time),
+                        )}
+                        active_course_count={data[i].active_course_count}
+                        quartile={data[i].total_activity_time_quartile}
+                        size={size}
+                    />
+                </td>,
+            );
+            aggregateActivityTime += Number(data[i].total_activity_time);
+            i += 1;
+        } else {
+            tableData.push(
+                <td className="block" key={dateCount.getTime()}>
+                    <Node date={new Date(dateCount)} size={size} quartile={0} />
+                </td>,
+            );
+        }
+        dateCount.setUTCDate(dateCount.getUTCDate() + 1);
+    }
+
+    useEffect(() => {
+        setAggregateActivityTime(Math.floor(aggregateActivityTime / 3600));
+    }, [aggregateActivityTime]);
+
+    /* add empty cells for days of week after activity range */
+    for (i = dateCount.getUTCDay(); i < 7; i++) {
+        tableData.push(
+            <td className="block" key={i + 7}>
+                <div className={`${nodeSizes[size]}`}></div>
+            </td>,
+        );
+    }
```

Comment:
do you think for this we could just add grey cells until the end of the year? I just was thinking this because its similar to how github does it. Open to what you think looks best though. @PThorpe92 do you have any thoughts / feelings towards this?
<img width="951" alt="Screenshot 2024-04-17 at 11 20 00 AM" src="https://github.com/UnlockedLabs/UnlockEdv2/assets/89868388/076f3d85-554a-46da-b26b-6e7df979345a">
<img width="832" alt="Screenshot 2024-04-17 at 11 21 30 AM" src="https://github.com/UnlockedLabs/UnlockEdv2/assets/89868388/833051bf-4bcb-4adb-bacc-2df4a4d683db">

Reply:
I like that :+1:  I think grey by default, so there is a year's worth at all times is the way to go.